File: apollo-runtime/src/main/java/com/apollographql/apollo/http/OkHttpExecutionContext.java
package com.apollographql.apollo.http;

import com.apollographql.apollo.api.ApolloExperimental;
import com.apollographql.apollo.api.ExecutionContext;
import okhttp3.Response;
import org.jetbrains.annotations.NotNull;

import static com.apollographql.apollo.api.internal.Utils.checkNotNull;

/**
 * Http GraphQL execution context, provides access to the raw {@link okhttp3.Response} response.
 */
@ApolloExperimental
public class OkHttpExecutionContext implements ExecutionContext.Element {

  public static final ExecutionContext.Key<OkHttpExecutionContext> KEY = new ExecutionContext.Key<OkHttpExecutionContext>() {
  };

  /**
   * Raw OkHttp response.
   */
  public final Response response;

  public OkHttpExecutionContext(@NotNull Response response) {
    checkNotNull(response, "response == null");
    this.response = strip(response);
  }

  @NotNull @Override public ExecutionContext.Key<?> getKey() {
    return KEY;
  }

  private static @NotNull Response strip(@NotNull Response response) {
    final Response.Builder builder = response.newBuilder();

    if (response.body() != null) {
      builder.body(null);
    }

    final Response cacheResponse = response.cacheResponse();
    if (cacheResponse != null) {
      builder.cacheResponse(strip(cacheResponse));
    }

    final Response networkResponse = response.networkResponse();
    if (networkResponse != null) {
      builder.networkResponse(strip(networkResponse));
    }

    return builder.build();
  }
}

File: apollo-runtime/src/main/java/com/apollographql/apollo3/http/OkHttpExecutionContext.kt
package com.apollographql.apollo3.http

import com.apollographql.apollo3.api.ApolloExperimental
import com.apollographql.apollo3.api.ExecutionContext
import com.apollographql.apollo3.api.ResponseContext
import okhttp3.Response

/**
 * Http GraphQL execution context, provides access to the raw {@link okhttp3.Response} response.
 */
@ApolloExperimental
class OkHttpExecutionContext(
    response: Response
) : ResponseContext(OkHttpExecutionContext) {
  companion object Key : ExecutionContext.Key<OkHttpExecutionContext>

  val response: Response = response.strip()

  private fun Response.strip(): Response {
    val builder = newBuilder()

    if (body() != null) {
      builder.body(null)
    }

    val cacheResponse = cacheResponse()
    if (cacheResponse != null) {
      builder.cacheResponse(cacheResponse.strip())
    }

    val networkResponse = networkResponse()
    if (networkResponse != null) {
      builder.networkResponse(networkResponse.strip())
    }

    return builder.build()
  }
}
